// egg-freelog-base/app/extend/restful-web-api/index.js
'use strict'

const restfulApiProd = require('./restful-apis-prod')
const restfulApiLocal = require('./restful-apis-local')

module.exports = class RestFulWebApis {

    constructor(config) {
        this.givenProvider = new restfulApiProd(config.gatewayUrl)
        // this.givenProvider = config.env === 'local'
        //     ? new restfulApiLocal(config.gatewayUrl)
        //     : new restfulApiProd(config.gatewayUrl)
    }

    /**
     * 用户信息
     */
    get userInfo() {
        return this.givenProvider.userInfo
    }

    /**
     * 资源信息
     * @returns {string}
     */
    get resourceInfo() {
        return this.givenProvider.resourceInfo
    }

    /**
     * 发行信息
     * @returns {string}
     */
    get releaseInfo() {
        return this.givenProvider.releaseInfo
    }

    /**
     * 获取presentable信息
     */
    get presentableInfo() {
        return this.givenProvider.presentableInfo
    }

    /**
     * 获取节点信息
     * @returns {string}
     */
    get nodeInfo() {
        return this.givenProvider.nodeInfo
    }

    /**
     * 测试节点
     * @returns {string}
     */
    get testNode(){
        return this.givenProvider.testNode
    }

    /**
     * 授权方案
     * @returns {string}
     */
    get authSchemeInfo() {
        return this.givenProvider.authSchemeInfo
    }

    /**
     * 合同信息
     * @returns {string}
     */
    get contractInfo() {
        return this.givenProvider.contractInfo
    }

    /**
     * 授权信息
     * @returns {string}
     */
    get authInfo() {
        return this.givenProvider.authInfo
    }

    /**
     * 用户/节点分组
     * @returns {string}
     */
    get groupInfo() {
        return this.givenProvider.groupInfo
    }

    /**
     * 统计计次服务
     * @returns {string}
     */
    get statisticsInfo() {
        return this.givenProvider.statisticsInfo
    }

    /**
     * 账户信息
     * @returns {string}
     */
    get accountInfo() {
        return this.givenProvider.accountInfo
    }

    /**
     * 支付信息
     * @returns {string}
     */
    get pay() {
        return this.givenProvider.pay
    }
}
