// egg-freelog-base/app/extend/restful-web-api/restful-apis-local.js
'use strict'

module.exports = class WebApiLocal {

    constructor(baseUrl) {
        this.baseUrl = baseUrl
    }

    /**
     * 用户信息
     * @returns {string}
     */
    get userInfo() {
        return 'http://127.0.0.1:7011/v1/userinfos'
    }

    /**
     * 资源信息
     * @returns {string}
     */
    get resourceInfo() {
        return 'http://127.0.0.1:7001/v1/resources'
    }

    /**
     * 获取presentable信息
     */
    get presentableInfo() {
        return 'http://127.0.0.1:7005/v1/presentables'
    }

    /**
     * 获取节点信息
     * @returns {string}
     */
    get nodeInfo() {
        return 'http://127.0.0.1:7005/v1/nodes'
    }

    /**
     * 测试节点
     */
    get testNode() {
        return 'http://127.0.0.1:7005/v1/testNodes'
    }

    /**
     * 授权方案
     * @returns {string}
     */
    get authSchemeInfo() {
        return 'http://127.0.0.1:7001/v1/resources/authSchemes'
    }

    /**
     * 发行信息
     * @returns {string}
     */
    get releaseInfo() {
        return 'http://127.0.0.1:7001/v1/releases'
    }

    /**
     * 合同信息
     * @returns {string}
     */
    get contractInfo() {
        return 'http://127.0.0.1:7008/v1/contracts'
    }

    /**
     * 授权信息
     * @returns {string}
     */
    get authInfo() {
        return 'http://127.0.0.1:7008/v1/auths'
    }

    /**
     * 用户/节点分组
     * @returns {string}
     */
    get groupInfo() {
        return 'http://127.0.0.1:7011/v1/groups'
    }

    /**
     * 统计计次服务
     * @returns {string}
     */
    get statisticsInfo() {
        return 'http://127.0.0.1:7018/v1/statistics'
    }

    /**
     * 账户信息
     * @returns {string}
     */
    get accountInfo() {
        return 'http://127.0.0.1:7055/v1/pay/accounts'
    }

    /**
     * 支付信息
     * @returns {string}
     */
    get pay() {
        return 'http://127.0.0.1:7055/v1/pay'
    }
}
