// egg-freelog-base/app/extend/helper/common_regex.js
/**
 * Created by yuliang on 2017/10/27.
 */

'use strict'

/**
 * 常用的freelog正则表达式验证
 * @type {{}}
 */
module.exports = {

    /**
     * 资源ID
     */
    resourceId: new RegExp(/^[0-9a-zA-Z]{40}$/),

    /**
     * 插件命名规则
     */
    widgetName: new RegExp(/^freelog-[a-z0-9._-]{3,15}-[a-z0-9._-]{2,14}[a-z0-9]{1}$/),

    /**
     * 分组ID
     */
    groupId: new RegExp(/^group_(node|user)_[0-9a-f]{24}$/),

    /**
     * 用户分组ID
     */
    userGroupId: new RegExp(/^group_user_[0-9a-f]{24}$/),

    /**
     * 节点分组ID
     */
    nodeGroupId: new RegExp(/^group_node_[0-9a-f]{24}$/),

    /**
     * md5
     */
    md5: new RegExp(/^[0-9a-f]{32}$/),

    /**
     * +86 地区移动手机号码
     */
    mobile86: new RegExp(/^1[34578]\d{9}$/),

    /**
     * email
     */
    email: new RegExp(/^[A-Za-z\d]+([-_.][A-Za-z\d]+)*@([A-Za-z\d]+[-.])+[A-Za-z\d]{2,4}$/),

    /**
     * mongodb-object-id
     */
    mongoObjectId: new RegExp(/^[0-9a-f]{24}$/),

    /**
     * 逗号分隔的resourceId
     */
    splitResourceId: new RegExp(/^[0-9a-zA-Z]{40}(,[0-9a-zA-Z]{40})*$/),

    /**
     * 逗号分隔的mongodbObjectId
     */
    splitMongoObjectId: new RegExp(/^[0-9a-f]{24}(,[0-9a-f]{24})*$/),

    /**
     * 逗号分隔的groupId
     */
    splitGroupId: new RegExp(/^group_(node|user)_[0-9a-f]{24}(,group_(node|user)_[0-9a-f]{24})*$/),

    /**
     * 逗号分隔的数字
     */
    splitNumber: new RegExp(/^\d+(,\d+)*$/),

    /**
     * 资源类型校验
     */
    resourceType: new RegExp(/^(?!_)[a-z0-9_]{3,20}(?<!_)$/),

    /**
     * 节点短域名
     */
    nodeDomain: new RegExp(/^(?!-)[a-z0-9-]{4,24}(?<!-)$/),

    /**
     * 节点名称(中文,英文,数字)
     */
    nodeName: new RegExp(/^[\u4E00-\u9FA5|a-zA-Z0-9]{4,20}$/),

    /**
     * 32位小写16进制
     */
    bit32Hex: new RegExp(/^[0-9a-f]{32}$/),

    /**
     * 交易账户ID
     */
    transferAccountId: new RegExp(/^(feth|fcny|fusd|feur)\d{3}[a-f0-9]{8}$/),

    /**
     * bucketName规则:只允许小写字母、数字、中划线（-），且不能以短横线开头或结尾
     */
    bucketName: new RegExp(/^(?!-)[a-z0-9-]{1,63}(?<!-)$/),

    /**
     * 字母、数字、中划线（-）组成,且不能以短横线开头或结尾,1-30长度
     */
    username: new RegExp(/^(?!-)[A-Za-z0-9-]{1,30}(?<!-)$/),

    /**
     * 发行(资源,mock,presentable统一)名称(不能包含\ / : * ? " < > |字符以及空格,长度1-60)
     */
    releaseName: new RegExp(/^(?!.*(\\|\/|:|\*|\?|"|<|>|\||\s|@|\$|#)).{1,60}$/),

    /**
     * 密码格式,6-24位任意字符,组合中必须包含数字和字母
     */
    password: new RegExp(/^(?=.*[0-9])(?=.*[a-zA-Z])(.{6,24})$/),

    /**
     * 发行全名规则,由username/releaseName组合而成
     */
    fullReleaseName: new RegExp(/^(?!-)[A-Za-z0-9-]{1,30}(?<!-)\/(?!.*(\\|\/|:|\*|\?|"|<|>|\||\s|@|\$|#)).{1,60}$/),
}
