// egg-freelog-base/app/enum/err_code.js
/**
 * Created by yuliang on 2017/8/9.
 */

'use strict'

/**
 * 二级错误码
 * @type {Object}
 */
module.exports = {

    //正常结果
    success: 0,

    //自动捕捉的错误
    autoSnapError: 1,

    //应用程序错误
    applicationError: 2,

    //授权错误
    authorizationError: 3,

    //参数相关的错误
    argumentError: 4,

    //API调用错误
    apiInvokingError: 5,

    //程序逻辑校验失败错误
    logicError: 6,

    //网络连接异常错误
    networkError: 7,

    //参数校验错误
    paramValidateError: 15,

    //oauth_client相关错误
    clientError: 16,

    //404
    notFoundPage: 17,

    //basic认证相关错误
    basicAuthError: 20,

    //用户密码验证相关错误
    passwordError: 21,

    //未找到代理代理服务器错误
    notFoundAgentServerError: 22,

    //请求超时错误
    requestTimeoutError: 23,

    //黑名单,禁止访问
    blackListError: 24,

    //白名单,没有权限
    whiteListError: 25,

    //源服务器请求异常
    originalServerError: 26,

    //token过期
    accessTokenTimeOutError: 27,

    //jtw错误
    jwtTokenAuthError: 28,

    //没有权限,被拒绝
    refuse: 29,

    //身份认证失败
    authenticationError: 30,

    //网关HTTP组件调用错误
    gatewayHttpComponentInvokingError: 31,

    //网关路由匹配错误
    gatewayRouterMatchError: 32,

    //网关上游API调用错误
    gatewayUpstreamApiError: 33,

    //异常占位区
    apiError: 100,
}
