// @freelog/resource-policy-lang/gen/expressionLexer.js
// Generated from expression.g4 by ANTLR 4.7.1
// jshint ignore: start
var antlr4 = require('antlr4/index');


var serializedATN = ["\u0003\u608b\ua72a\u8133\ub9ed\u417c\u3be7\u7786\u5964",
    "\u0002\u0015\u00e2\b\u0001\u0004\u0002\t\u0002\u0004\u0003\t\u0003\u0004",
    "\u0004\t\u0004\u0004\u0005\t\u0005\u0004\u0006\t\u0006\u0004\u0007\t",
    "\u0007\u0004\b\t\b\u0004\t\t\t\u0004\n\t\n\u0004\u000b\t\u000b\u0004",
    "\f\t\f\u0004\r\t\r\u0004\u000e\t\u000e\u0004\u000f\t\u000f\u0004\u0010",
    "\t\u0010\u0004\u0011\t\u0011\u0004\u0012\t\u0012\u0004\u0013\t\u0013",
    "\u0004\u0014\t\u0014\u0004\u0015\t\u0015\u0004\u0016\t\u0016\u0004\u0017",
    "\t\u0017\u0004\u0018\t\u0018\u0004\u0019\t\u0019\u0004\u001a\t\u001a",
    "\u0004\u001b\t\u001b\u0004\u001c\t\u001c\u0004\u001d\t\u001d\u0004\u001e",
    "\t\u001e\u0004\u001f\t\u001f\u0004 \t \u0004!\t!\u0004\"\t\"\u0004#",
    "\t#\u0004$\t$\u0004%\t%\u0004&\t&\u0004\'\t\'\u0004(\t(\u0004)\t)\u0004",
    "*\t*\u0004+\t+\u0004,\t,\u0004-\t-\u0004.\t.\u0004/\t/\u00040\t0\u0003",
    "\u0002\u0003\u0002\u0003\u0003\u0003\u0003\u0003\u0004\u0003\u0004\u0003",
    "\u0005\u0003\u0005\u0003\u0006\u0003\u0006\u0003\u0007\u0003\u0007\u0003",
    "\b\u0003\b\u0003\t\u0003\t\u0003\n\u0003\n\u0003\u000b\u0003\u000b\u0003",
    "\f\u0003\f\u0003\r\u0003\r\u0003\u000e\u0003\u000e\u0003\u000f\u0003",
    "\u000f\u0003\u0010\u0003\u0010\u0003\u0011\u0003\u0011\u0003\u0012\u0003",
    "\u0012\u0003\u0013\u0003\u0013\u0003\u0014\u0003\u0014\u0003\u0015\u0003",
    "\u0015\u0003\u0016\u0003\u0016\u0003\u0017\u0003\u0017\u0003\u0018\u0003",
    "\u0018\u0003\u0019\u0003\u0019\u0003\u001a\u0003\u001a\u0003\u001b\u0003",
    "\u001b\u0003\u001c\u0006\u001c\u0097\n\u001c\r\u001c\u000e\u001c\u0098",
    "\u0003\u001d\u0005\u001d\u009c\n\u001d\u0003\u001d\u0003\u001d\u0003",
    "\u001e\u0006\u001e\u00a1\n\u001e\r\u001e\u000e\u001e\u00a2\u0003\u001e",
    "\u0007\u001e\u00a6\n\u001e\f\u001e\u000e\u001e\u00a9\u000b\u001e\u0003",
    "\u001f\u0003\u001f\u0003 \u0003 \u0003!\u0003!\u0003\"\u0003\"\u0003",
    "#\u0003#\u0003$\u0003$\u0003%\u0003%\u0003&\u0003&\u0003\'\u0003\'\u0003",
    "(\u0003(\u0003)\u0003)\u0003*\u0003*\u0003+\u0003+\u0003+\u0003,\u0003",
    ",\u0003,\u0005,\u00c9\n,\u0003,\u0005,\u00cc\n,\u0003-\u0003-\u0003",
    ".\u0003.\u0003.\u0003.\u0003/\u0006/\u00d5\n/\r/\u000e/\u00d6\u0003",
    "/\u0003/\u0006/\u00db\n/\r/\u000e/\u00dc\u0005/\u00df\n/\u00030\u0003",
    "0\u0002\u00021\u0003\u0002\u0005\u0002\u0007\u0002\t\u0002\u000b\u0002",
    "\r\u0002\u000f\u0002\u0011\u0002\u0013\u0002\u0015\u0002\u0017\u0002",
    "\u0019\u0002\u001b\u0002\u001d\u0002\u001f\u0002!\u0002#\u0002%\u0002",
    "\'\u0002)\u0002+\u0002-\u0002/\u00021\u00023\u00025\u00027\u00039\u0004",
    ";\u0005=\u0006?\u0007A\bC\tE\nG\u000bI\fK\rM\u000eO\u000fQ\u0010S\u0011",
    "U\u0012W\u0013Y\u0014[\u0015]\u0002_\u0002\u0003\u0002 \u0004\u0002",
    "CCcc\u0004\u0002DDdd\u0004\u0002EEee\u0004\u0002FFff\u0004\u0002GGg",
    "g\u0004\u0002HHhh\u0004\u0002IIii\u0004\u0002JJjj\u0004\u0002KKkk\u0004",
    "\u0002LLll\u0004\u0002MMmm\u0004\u0002NNnn\u0004\u0002OOoo\u0004\u0002",
    "PPpp\u0004\u0002QQqq\u0004\u0002RRrr\u0004\u0002SSss\u0004\u0002TTt",
    "t\u0004\u0002UUuu\u0004\u0002VVvv\u0004\u0002WWww\u0004\u0002XXxx\u0004",
    "\u0002YYyy\u0004\u0002ZZzz\u0004\u0002[[{{\u0004\u0002\\\\||\u0003\u0002",
    "2;\u0004\u0002C\\c|\u0006\u00022;C\\aac|\u0004\u0002--//\u0002\u00ce",
    "\u00027\u0003\u0002\u0002\u0002\u00029\u0003\u0002\u0002\u0002\u0002",
    ";\u0003\u0002\u0002\u0002\u0002=\u0003\u0002\u0002\u0002\u0002?\u0003",
    "\u0002\u0002\u0002\u0002A\u0003\u0002\u0002\u0002\u0002C\u0003\u0002",
    "\u0002\u0002\u0002E\u0003\u0002\u0002\u0002\u0002G\u0003\u0002\u0002",
    "\u0002\u0002I\u0003\u0002\u0002\u0002\u0002K\u0003\u0002\u0002\u0002",
    "\u0002M\u0003\u0002\u0002\u0002\u0002O\u0003\u0002\u0002\u0002\u0002",
    "Q\u0003\u0002\u0002\u0002\u0002S\u0003\u0002\u0002\u0002\u0002U\u0003",
    "\u0002\u0002\u0002\u0002W\u0003\u0002\u0002\u0002\u0002Y\u0003\u0002",
    "\u0002\u0002\u0002[\u0003\u0002\u0002\u0002\u0003a\u0003\u0002\u0002",
    "\u0002\u0005c\u0003\u0002\u0002\u0002\u0007e\u0003\u0002\u0002\u0002",
    "\tg\u0003\u0002\u0002\u0002\u000bi\u0003\u0002\u0002\u0002\rk\u0003",
    "\u0002\u0002\u0002\u000fm\u0003\u0002\u0002\u0002\u0011o\u0003\u0002",
    "\u0002\u0002\u0013q\u0003\u0002\u0002\u0002\u0015s\u0003\u0002\u0002",
    "\u0002\u0017u\u0003\u0002\u0002\u0002\u0019w\u0003\u0002\u0002\u0002",
    "\u001by\u0003\u0002\u0002\u0002\u001d{\u0003\u0002\u0002\u0002\u001f",
    "}\u0003\u0002\u0002\u0002!\u007f\u0003\u0002\u0002\u0002#\u0081\u0003",
    "\u0002\u0002\u0002%\u0083\u0003\u0002\u0002\u0002\'\u0085\u0003\u0002",
    "\u0002\u0002)\u0087\u0003\u0002\u0002\u0002+\u0089\u0003\u0002\u0002",
    "\u0002-\u008b\u0003\u0002\u0002\u0002/\u008d\u0003\u0002\u0002\u0002",
    "1\u008f\u0003\u0002\u0002\u00023\u0091\u0003\u0002\u0002\u00025\u0093",
    "\u0003\u0002\u0002\u00027\u0096\u0003\u0002\u0002\u00029\u009b\u0003",
    "\u0002\u0002\u0002;\u00a0\u0003\u0002\u0002\u0002=\u00aa\u0003\u0002",
    "\u0002\u0002?\u00ac\u0003\u0002\u0002\u0002A\u00ae\u0003\u0002\u0002",
    "\u0002C\u00b0\u0003\u0002\u0002\u0002E\u00b2\u0003\u0002\u0002\u0002",
    "G\u00b4\u0003\u0002\u0002\u0002I\u00b6\u0003\u0002\u0002\u0002K\u00b8",
    "\u0003\u0002\u0002\u0002M\u00ba\u0003\u0002\u0002\u0002O\u00bc\u0003",
    "\u0002\u0002\u0002Q\u00be\u0003\u0002\u0002\u0002S\u00c0\u0003\u0002",
    "\u0002\u0002U\u00c2\u0003\u0002\u0002\u0002W\u00c5\u0003\u0002\u0002",
    "\u0002Y\u00cd\u0003\u0002\u0002\u0002[\u00cf\u0003\u0002\u0002\u0002",
    "]\u00d4\u0003\u0002\u0002\u0002_\u00e0\u0003\u0002\u0002\u0002ab\t\u0002",
    "\u0002\u0002b\u0004\u0003\u0002\u0002\u0002cd\t\u0003\u0002\u0002d\u0006",
    "\u0003\u0002\u0002\u0002ef\t\u0004\u0002\u0002f\b\u0003\u0002\u0002",
    "\u0002gh\t\u0005\u0002\u0002h\n\u0003\u0002\u0002\u0002ij\t\u0006\u0002",
    "\u0002j\f\u0003\u0002\u0002\u0002kl\t\u0007\u0002\u0002l\u000e\u0003",
    "\u0002\u0002\u0002mn\t\b\u0002\u0002n\u0010\u0003\u0002\u0002\u0002",
    "op\t\t\u0002\u0002p\u0012\u0003\u0002\u0002\u0002qr\t\n\u0002\u0002",
    "r\u0014\u0003\u0002\u0002\u0002st\t\u000b\u0002\u0002t\u0016\u0003\u0002",
    "\u0002\u0002uv\t\f\u0002\u0002v\u0018\u0003\u0002\u0002\u0002wx\t\r",
    "\u0002\u0002x\u001a\u0003\u0002\u0002\u0002yz\t\u000e\u0002\u0002z\u001c",
    "\u0003\u0002\u0002\u0002{|\t\u000f\u0002\u0002|\u001e\u0003\u0002\u0002",
    "\u0002}~\t\u0010\u0002\u0002~ \u0003\u0002\u0002\u0002\u007f\u0080\t",
    "\u0011\u0002\u0002\u0080\"\u0003\u0002\u0002\u0002\u0081\u0082\t\u0012",
    "\u0002\u0002\u0082$\u0003\u0002\u0002\u0002\u0083\u0084\t\u0013\u0002",
    "\u0002\u0084&\u0003\u0002\u0002\u0002\u0085\u0086\t\u0014\u0002\u0002",
    "\u0086(\u0003\u0002\u0002\u0002\u0087\u0088\t\u0015\u0002\u0002\u0088",
    "*\u0003\u0002\u0002\u0002\u0089\u008a\t\u0016\u0002\u0002\u008a,\u0003",
    "\u0002\u0002\u0002\u008b\u008c\t\u0017\u0002\u0002\u008c.\u0003\u0002",
    "\u0002\u0002\u008d\u008e\t\u0018\u0002\u0002\u008e0\u0003\u0002\u0002",
    "\u0002\u008f\u0090\t\u0019\u0002\u0002\u00902\u0003\u0002\u0002\u0002",
    "\u0091\u0092\t\u001a\u0002\u0002\u00924\u0003\u0002\u0002\u0002\u0093",
    "\u0094\t\u001b\u0002\u0002\u00946\u0003\u0002\u0002\u0002\u0095\u0097",
    "\t\u001c\u0002\u0002\u0096\u0095\u0003\u0002\u0002\u0002\u0097\u0098",
    "\u0003\u0002\u0002\u0002\u0098\u0096\u0003\u0002\u0002\u0002\u0098\u0099",
    "\u0003\u0002\u0002\u0002\u00998\u0003\u0002\u0002\u0002\u009a\u009c",
    "\u0007\u000f\u0002\u0002\u009b\u009a\u0003\u0002\u0002\u0002\u009b\u009c",
    "\u0003\u0002\u0002\u0002\u009c\u009d\u0003\u0002\u0002\u0002\u009d\u009e",
    "\u0007\f\u0002\u0002\u009e:\u0003\u0002\u0002\u0002\u009f\u00a1\t\u001d",
    "\u0002\u0002\u00a0\u009f\u0003\u0002\u0002\u0002\u00a1\u00a2\u0003\u0002",
    "\u0002\u0002\u00a2\u00a0\u0003\u0002\u0002\u0002\u00a2\u00a3\u0003\u0002",
    "\u0002\u0002\u00a3\u00a7\u0003\u0002\u0002\u0002\u00a4\u00a6\t\u001e",
    "\u0002\u0002\u00a5\u00a4\u0003\u0002\u0002\u0002\u00a6\u00a9\u0003\u0002",
    "\u0002\u0002\u00a7\u00a5\u0003\u0002\u0002\u0002\u00a7\u00a8\u0003\u0002",
    "\u0002\u0002\u00a8<\u0003\u0002\u0002\u0002\u00a9\u00a7\u0003\u0002",
    "\u0002\u0002\u00aa\u00ab\u0007*\u0002\u0002\u00ab>\u0003\u0002\u0002",
    "\u0002\u00ac\u00ad\u0007+\u0002\u0002\u00ad@\u0003\u0002\u0002\u0002",
    "\u00ae\u00af\u0007-\u0002\u0002\u00afB\u0003\u0002\u0002\u0002\u00b0",
    "\u00b1\u0007/\u0002\u0002\u00b1D\u0003\u0002\u0002\u0002\u00b2\u00b3",
    "\u0007,\u0002\u0002\u00b3F\u0003\u0002\u0002\u0002\u00b4\u00b5\u0007",
    "1\u0002\u0002\u00b5H\u0003\u0002\u0002\u0002\u00b6\u00b7\u0007@\u0002",
    "\u0002\u00b7J\u0003\u0002\u0002\u0002\u00b8\u00b9\u0007>\u0002\u0002",
    "\u00b9L\u0003\u0002\u0002\u0002\u00ba\u00bb\u0007?\u0002\u0002\u00bb",
    "N\u0003\u0002\u0002\u0002\u00bc\u00bd\u0007.\u0002\u0002\u00bdP\u0003",
    "\u0002\u0002\u0002\u00be\u00bf\u00070\u0002\u0002\u00bfR\u0003\u0002",
    "\u0002\u0002\u00c0\u00c1\u0007`\u0002\u0002\u00c1T\u0003\u0002\u0002",
    "\u0002\u00c2\u00c3\u0007r\u0002\u0002\u00c3\u00c4\u0007k\u0002\u0002",
    "\u00c4V\u0003\u0002\u0002\u0002\u00c5\u00cb\u0005]/\u0002\u00c6\u00c8",
    "\t\u0006\u0002\u0002\u00c7\u00c9\u0005_0\u0002\u00c8\u00c7\u0003\u0002",
    "\u0002\u0002\u00c8\u00c9\u0003\u0002\u0002\u0002\u00c9\u00ca\u0003\u0002",
    "\u0002\u0002\u00ca\u00cc\u0005]/\u0002\u00cb\u00c6\u0003\u0002\u0002",
    "\u0002\u00cb\u00cc\u0003\u0002\u0002\u0002\u00ccX\u0003\u0002\u0002",
    "\u0002\u00cd\u00ce\u0007g\u0002\u0002\u00ceZ\u0003\u0002\u0002\u0002",
    "\u00cf\u00d0\u0007u\u0002\u0002\u00d0\u00d1\u0007w\u0002\u0002\u00d1",
    "\u00d2\u0007o\u0002\u0002\u00d2\\\u0003\u0002\u0002\u0002\u00d3\u00d5",
    "\u00042;\u0002\u00d4\u00d3\u0003\u0002\u0002\u0002\u00d5\u00d6\u0003",
    "\u0002\u0002\u0002\u00d6\u00d4\u0003\u0002\u0002\u0002\u00d6\u00d7\u0003",
    "\u0002\u0002\u0002\u00d7\u00de\u0003\u0002\u0002\u0002\u00d8\u00da\u0007",
    "0\u0002\u0002\u00d9\u00db\u00042;\u0002\u00da\u00d9\u0003\u0002\u0002",
    "\u0002\u00db\u00dc\u0003\u0002\u0002\u0002\u00dc\u00da\u0003\u0002\u0002",
    "\u0002\u00dc\u00dd\u0003\u0002\u0002\u0002\u00dd\u00df\u0003\u0002\u0002",
    "\u0002\u00de\u00d8\u0003\u0002\u0002\u0002\u00de\u00df\u0003\u0002\u0002",
    "\u0002\u00df^\u0003\u0002\u0002\u0002\u00e0\u00e1\t\u001f\u0002\u0002",
    "\u00e1`\u0003\u0002\u0002\u0002\f\u0002\u0098\u009b\u00a2\u00a7\u00c8",
    "\u00cb\u00d6\u00dc\u00de\u0002"].join("");


var atn = new antlr4.atn.ATNDeserializer().deserialize(serializedATN);

var decisionsToDFA = atn.decisionToState.map( function(ds, index) { return new antlr4.dfa.DFA(ds, index); });

function expressionLexer(input) {
	antlr4.Lexer.call(this, input);
    this._interp = new antlr4.atn.LexerATNSimulator(this, atn, decisionsToDFA, new antlr4.PredictionContextCache());
    return this;
}

expressionLexer.prototype = Object.create(antlr4.Lexer.prototype);
expressionLexer.prototype.constructor = expressionLexer;

Object.defineProperty(expressionLexer.prototype, "atn", {
        get : function() {
                return atn;
        }
});

expressionLexer.EOF = antlr4.Token.EOF;
expressionLexer.INT = 1;
expressionLexer.NEWLINE = 2;
expressionLexer.VARIABLE = 3;
expressionLexer.LPAREN = 4;
expressionLexer.RPAREN = 5;
expressionLexer.PLUS = 6;
expressionLexer.MINUS = 7;
expressionLexer.TIMES = 8;
expressionLexer.DIV = 9;
expressionLexer.GT = 10;
expressionLexer.LT = 11;
expressionLexer.EQ = 12;
expressionLexer.COMMA = 13;
expressionLexer.POINT = 14;
expressionLexer.POW = 15;
expressionLexer.PI = 16;
expressionLexer.SCIENTIFIC_NUMBER = 17;
expressionLexer.EULER = 18;
expressionLexer.SUM = 19;

expressionLexer.prototype.channelNames = [ "DEFAULT_TOKEN_CHANNEL", "HIDDEN" ];

expressionLexer.prototype.modeNames = [ "DEFAULT_MODE" ];

expressionLexer.prototype.literalNames = [ null, null, null, null, "'('", 
                                           "')'", "'+'", "'-'", "'*'", "'/'", 
                                           "'>'", "'<'", "'='", "','", "'.'", 
                                           "'^'", "'pi'", null, "'e'", "'sum'" ];

expressionLexer.prototype.symbolicNames = [ null, "INT", "NEWLINE", "VARIABLE", 
                                            "LPAREN", "RPAREN", "PLUS", 
                                            "MINUS", "TIMES", "DIV", "GT", 
                                            "LT", "EQ", "COMMA", "POINT", 
                                            "POW", "PI", "SCIENTIFIC_NUMBER", 
                                            "EULER", "SUM" ];

expressionLexer.prototype.ruleNames = [ "A", "B", "C", "D", "E", "F", "G", 
                                        "H", "I", "J", "K", "L", "M", "N", 
                                        "O", "P", "Q", "R", "S", "T", "U", 
                                        "V", "W", "X", "Y", "Z", "INT", 
                                        "NEWLINE", "VARIABLE", "LPAREN", 
                                        "RPAREN", "PLUS", "MINUS", "TIMES", 
                                        "DIV", "GT", "LT", "EQ", "COMMA", 
                                        "POINT", "POW", "PI", "SCIENTIFIC_NUMBER", 
                                        "EULER", "SUM", "NUMBER", "SIGN" ];

expressionLexer.prototype.grammarFileName = "expression.g4";



exports.expressionLexer = expressionLexer;
