// egg-freelog-base/app/extend/file-oss/aliyun-oss.js
/**
 * Created by yuliang on 2017/7/4.
 */

'use strict'

const lodash = require('lodash')

module.exports = class aliOssUpload {

    constructor(config) {
        const aliOss = require('ali-oss')
        const wrapper = aliOss.Wrapper || aliOss
        this.client = new wrapper(config)
    }

    /**
     * 以buffer的形式上传文件到阿里云OSS
     * WIKI:https://help.aliyun.com/document_detail/32072.html?spm=5176.doc32070.6.754.A49ih2
     * @param filePath 文件上传路径
     * @param fileBuffer 文件buffer
     * @returns {Promise.<T>}
     */
    putBuffer(filePath, fileBuffer, options) {
        return this.client.put(filePath, fileBuffer, options)
    }

    /**
     * 以stream的形式上传文件到阿里云OSS
     * WIKI:https://help.aliyun.com/document_detail/32072.html?spm=5176.doc32070.6.754.A49ih2
     * @param filePath 文件上传路径
     * @param fileStream 文件流
     * @param contentLength 文件大小
     * @returns {Promise.<T>}
     */
    putStream(filePath, fileStream, options) {
        return this.client.putStream(filePath, fileStream, options)
    }

    /**
     * 删除文件
     * @param objectKey
     * @returns {void | Router | IterableIterator<*> | * | boolean | Promise<boolean>}
     */
    deleteFile(objectKey) {
        return this.client.delete(objectKey)
    }

    /**
     * 复制文件
     * @param to 目标objectKey
     * @param from 来源objectKey
     * @param options 可选,meta覆盖
     * @returns {*|IterableIterator<*|undefined>|number}
     */
    copyFile(to, from, options) {
        return this.client.copy(to, from, options)
    }
}
